### PR TITLE
gh-67413: Fix segfaults and multiple leaks in the lzma and bz2 modules

### PR DESCRIPTION
**Pull request by ZackerySpytz**
lzma.LZMADecompressor and bz2.BZ2Decompressor objects caused
segfaults when their `__init__()` methods were not called.

lzma.LZMADecompressor, lzma.LZMACompressor, bz2.BZ2Compressor,
and bz2.BZ2Decompressor objects would leak locks and internal buffers
when their `__init__()` methods were called multiple times.

https://bugs.python.org/issue23224

Automerge-Triggered-By: GH:erlend-aasland

* Issue: gh-67413

### CODE REVIEW

**Comment by ZackerySpytz** [comment]
It seems that this PR will also fix the issue described in bpo-33916.

**Comment by vstinner** [comment]
Hello @ZackerySpytz, since I merged my PR #7843, would you mind to rebase your change please? (merge master into your PR)

**Comment by ZackerySpytz** [comment]
@vstinner Done.

File: Lib/test/test_bz2.py [hunk 1]
```diff
@@ -826,6 +826,10 @@ def test_refleaks_in___init__(self):
             bzd.__init__()
         self.assertAlmostEqual(gettotalrefcount() - refs_before, 0, delta=10)
 
+    def test_uninitialized_BZ2Decompressor_crash(self):
```
Comment (izbyshev) [inline Lib/test/test_bz2.py:847]: Should tests for `BZ2Compressor` and `LZMACompressor` be added too?

File: Misc/NEWS.d/next/Library/2018-06-20-09-12-21.bpo-23224.zxCQ13.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Fix segfaults when creating :class:`lzma.LZMADecompressor` and
+:class:`bz2.BZ2Decompressor` objects without calling ``__init__()``, and fix
```
Comment (izbyshev) [inline Misc/NEWS.d/next/Library/2018-06-20-09-12-21.bpo-23224.zxCQ13.rst:2]: Segfaults apply for `*Compressor` objects too.

**Comment by csabella** [comment]
@ZackerySpytz, please rebase to resolve conflicts.  Also, please address @izbyshev's review.  Thanks!

File: Misc/NEWS.d/next/Library/2018-06-20-09-12-21.bpo-23224.zxCQ13.rst (hunk 2 above)
Comment (ZackerySpytz) [reply Misc/NEWS.d/next/Library/2018-06-20-09-12-21.bpo-23224.zxCQ13.rst:2]: @izbyshev I don't think so. Please explain why you think segfaults can occur with `*Compressor` objects.

**Comment by csabella** [comment]
@ZackerySpytz ping

File: Misc/NEWS.d/next/Library/2018-06-20-09-12-21.bpo-23224.zxCQ13.rst (hunk 2 above)
Comment (izbyshev) [reply Misc/NEWS.d/next/Library/2018-06-20-09-12-21.bpo-23224.zxCQ13.rst:2]: An uninitialized `*Compressor` object will crash if any of its methods is called:
```
from bz2 import BZ2Compressor as C
C.__new__(C).compress(b'')
Segmentation fault (core dumped)
```

**Comment by gpshead** [comment]
could this PR be updated?

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]
There are merge conflicts.

**Comment by erlend-aasland** [comment]
(Removed Victor from the review list, as he's off CPython these days.)

File: Misc/NEWS.d/next/Library/2018-06-20-09-12-21.bpo-23224.zxCQ13.rst (hunk 2 above)
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2018-06-20-09-12-21.bpo-23224.zxCQ13.rst:2]: Post this PR:

```python
Python 3.12.0a5+ (heads/bpo-23224-lzma-bz2-segfault-dirty:66dc4eb5f9, Feb 23 2023, 00:13:14) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from bz2 import BZ2Compressor as C
>>> C.__new__(C).compress(b'')
b''
```

Marking this as resolved.

**Review by kumaraditya303 — APPROVED** [review]

**Comment by erlend-aasland** [comment]
@kumaraditya303: thanks for the review. I did a small PEP-7 change to the _added_ code in aad76e994253f32a6ed9e8c1e89018a32de080c7. Will land when the CI is green.

**Comment by erlend-aasland** [comment]
Oh, Miss Islington was faster than GitHub; aad76e9 is now in limbo :)

**Comment by kumaraditya303** [comment]
> Oh, Miss Islington was faster than GitHub; https://github.com/python/cpython/commit/aad76e994253f32a6ed9e8c1e89018a32de080c7 is now in limbo :)

"The missing piece of puzzle all come together when you least expect it." :)

**Comment by vstinner** [comment]
Better late than never! Thanks @ZackerySpytz for the initial fix and @erlend-aasland for the updates.

**Pull request merged**